apiserver/admin.go

```diff
@@ -421,14 +420,38 @@ func (f modelUserEntityFinder) FindEntity(tag names.Tag) (state.Entity, error) {
 		return f.st.FindEntity(tag)
 	}
 
-	modelUser, controllerUser, err := common.UserAccess(f.st, utag)
-	if err != nil {
+	modelUser, err := f.st.UserAccess(utag, f.st.ModelTag())
+	if err != nil && !errors.IsNotFound(err) {
 		return nil, errors.Trace(err)
 	}
+	// No model user found, so see if the user has been granted
+	// access to the controller.
+	if permission.IsEmptyUserAccess(modelUser) {
+		controllerUser, err := state.ControllerAccess(f.st, utag)
+		if err != nil && !errors.IsNotFound(err) {
+			return nil, errors.Trace(err)
+		}
+		// TODO(perrito666) remove the following section about everyone group
+		// when groups are implemented, this accounts only for the lack of a local
+		// ControllerUser when logging in from an external user that has not been granted
+		// permissions on the controller but there are permissions for the special
+		// everyone group.
+		if permission.IsEmptyUserAccess(controllerUser) && !utag.IsLocal() {
+			everyoneTag := names.NewUserTag(common.EveryoneTagName)
+			controllerUser, err = f.st.UserAccess(everyoneTag, f.st.ControllerTag())
+			if err != nil && !errors.IsNotFound(err) {
+				return nil, errors.Annotatef(err, "obtaining ControllerUser for everyone group")
+			}
+		}
+		if permission.IsEmptyUserAccess(controllerUser) {
+			return nil, errors.NotFoundf("model or controller user")
+		}
+	}
+
 	u := &modelUserEntity{
-		st:             f.st,
-		modelUser:      modelUser,
-		controllerUser: controllerUser,
+		st:        f.st,
+		modelUser: modelUser,
```

Comment:
The only place that used controllerUser was to call Tag() and get the controller tag. So I refactored how the tag was set and thus the controllerUser can be deleted.